### PR TITLE
Code nav: single-click jump-to-definition

### PR DESCRIPTION
**Pull request by chrismwendt**
This lets you jump to definition by clicking on the token:

You can still select text even though the cursor is the hand icon 👆

This is a prototype and the plan is to merge it behind a feature flag to get a feel for the UX.

**Motivation**

This is similar to how [Bazel](https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/util/SimpleLogHandler.java;l=921;drc=d888e4d44fe1eb04a7122ebb206b2255e6862f1a;bpv=;bpt=1) and [Haskell Code Explorer](https://haskell-code-explorer.mfix.io/package/stm-2.4.5.0/show/Control/Concurrent/STM/TQueue.hs#L108) work and was [requested by a user](https://sourcegraph.slack.com/archives/C0W2E592M/p1552492262046200).

**Feature flag**

Global/org/user settings:

```
  "singleClickGoToDefinition": true,
```

TODO

- [ ] Merge https://github.com/sourcegraph/codeintellify/pull/106 (depends on a new API)
- [x] Figure out how to avoid needing to assign to `window.location` (currently causes a full page load when jumping to a different file)
- [x] Put this behind a feature flag

See proposal doc https://sourcegraph.slack.com/archives/CHXHX7XAS/p1555110738009200

### CODE REVIEW

**Comment by ijsnow** [comment]
I worry that this will add more confusion around the ability to select text under the cursor. We had a lot of complaints from users who thought we were taking away the ability to select text and this seems like it will actually interfere. For example I usually double click tokens that I want to copy because that automatically selects the whole word. If I do that with this enabled, will it navigate me or select the text?

**Comment by ijsnow** [comment]
Also what will be the UX for creating links with a position so that a tooltip will open for the person following the link? [Here's an example when I found this to be very useful.](https://github.com/sourcegraph/sourcegraph/pull/3337#discussion_r275996972)

**Comment by chrismwendt** [comment]
> For example I usually double click tokens that I want to copy because that automatically selects the whole word. If I do that with this enabled, will it navigate me or select the text?

As this PR stands right now, the first click of a double click does not trigger navigation, so double clicking doesn't work. A small debounce could be added to support double clicking at the expense of having to wait a few hundred milliseconds extra on each jump-to-definition.

> Also what will be the UX for creating links with a position so that a tooltip will open for the person following the link? Here's an example when I found this to be very useful.

As this PR stands right now, deep links will not trigger a hover at all. Perhaps you can express what auto-pinning enabled you to do and explain why it's an important use case.

File: web/src/repo/blob/Blob.tsx [hunk 1]
```diff
@@ -197,8 +200,27 @@ export class Blob extends React.Component<BlobProps, BlobState> {
                 dom: domFunctions,
             })
         )
+        const j2d = (ev: MouseEvent) => {
```
Comment (lguychard) [inline web/src/repo/blob/Blob.tsx]: For readability:

```suggestion
        const jumpToDefinition = (ev: MouseEvent) => {
```

What's more, our codebase currently uses `goToDefinition`, not `jumpToDefinition`, any reason to break  consistency here (and go with `singleClickGoToDefinition`, `goToDefinition`, `goToDefinitionAction`)?

File: web/src/repo/blob/Blob.tsx [hunk 2]
```diff
@@ -197,8 +200,27 @@ export class Blob extends React.Component<BlobProps, BlobState> {
...
+                this.state.actionsOrError instanceof Array &&
+                this.state.actionsOrError.find(a => a.action.id === 'goToDefinition.preloaded')
+            if (j2daction) {
+                window.location = j2daction.action.commandArguments![0]
+                ev.stopPropagation()
+            }
+        }
         this.subscriptions.add(
             hoverifier.hoverStateUpdates.subscribe(update => {
+                if (singleClickJ2D && this.state.token !== update.token) {
```
Comment (lguychard) [inline web/src/repo/blob/Blob.tsx]: Rel. https://github.com/sourcegraph/codeintellify/pull/106#discussion_r276185098, from naming It's really counter-intuitive to me that `token` is a HTML element.

**Comment by ijsnow** [comment]
> so double clicking doesn't work

:/

> Perhaps you can express what auto-pinning enabled you to do and explain why it's an important use case

It is often very useful to have the hover open automatically from a link. 

- Sending links to a specific symbol to someone you are collaborating with
- Symbol search

Without the tooltip automatically opening, the user has no context of what they are supposed to be looking at. For the case of clicking on a symbol search result, it will add overhead for the user to find the token that they just decided they wanted to see from the search result.

**Comment by drenehtsral** [comment]
I was one of the people originally wanting single-click navigation, but letting users configure their UI preferences is probably the most important.  This thread illustrates well that there will never be a one-size-fits-all UI for something like this and the more in-your-face the UI is the more visceral the reaction is.

This is definitely better than relying on a hover-over tool-tip since that excludes both keyboard-only users and touch-screen users.  I imagine I'm in the minority but I'd still prefer a plain hypertext interface (no client-side scripts or dynamic elements).

File: web/src/repo/blob/Blob.tsx (hunk 1 above)
Comment (chrismwendt) [reply web/src/repo/blob/Blob.tsx]: :+1: Renamed to `goToDefinition` et al.

**Comment by chrismwendt** [comment]
@lguychard Did you want to do another review?

File: web/src/repo/blob/Blob.tsx [hunk 3]
```diff
@@ -197,8 +204,27 @@ export class Blob extends React.Component<BlobProps, BlobState> {
                 dom: domFunctions,
             })
         )
+        const goToDefinition = (ev: MouseEvent) => {
+            const goToDefinitionAction =
+                this.state.actionsOrError instanceof Array &&
+                this.state.actionsOrError.find(action => action.action.id === 'goToDefinition.preloaded')
```
Comment (felixfbecker) [inline web/src/repo/blob/Blob.tsx]: This looks hacky, and like a circumvention/break of the `getActions()` and Action abstraction

Comment (chrismwendt) [reply web/src/repo/blob/Blob.tsx]: Perhaps we could introduce another concept like a "primary action" which would be assigned to single-click and defaults to `goToDefinition`. WDYT?

File: web/src/repo/blob/Blob.tsx [hunk 4]
```diff
@@ -197,8 +204,27 @@ export class Blob extends React.Component<BlobProps, BlobState> {
...
+            if (goToDefinitionAction) {
+                this.props.history.push(goToDefinitionAction.action.commandArguments![0])
+                ev.stopPropagation()
+            }
+        }
         this.subscriptions.add(
             hoverifier.hoverStateUpdates.subscribe(update => {
+                if (singleClickGoToDefinition && this.state.hoveredTokenElement !== update.hoveredTokenElement) {
+                    if (this.state.hoveredTokenElement) {
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply web/src/repo/blob/Blob.tsx]: Oh, I see! How about one of these options:

```typescript
// Current state of PR
updates.subscribe(update => {
  if (state.element ) { state.element.removeListener() }
  if (update.element) { update.element.addListener()   }
  setState(update)
})

// Option 1: mutable variable
let element
updates.subscribe(update => {
  if (element       ) { element.removeListener()     }
  if (update.element) { update.element.addListener() }
  element = update.element
})

// Option 2: `scan`
update.pipe(
  scan(
    (element, update) => {
      if (element       ) { element.removeListener()     }
      if (update.element) { update.element.addListener() }
      return update.element
    },
    undefined)
).subscribe()

// Option 3: `pairwise`
update.pipe(map(update => update.elemnt), pairwise()).subscribe(
  ([current, next]) => {
    if (current) { current.removeListener() }
    if (next   ) { next.addListener()       }
    return update
  }
)
```

[the remark this answers is not in the document]
Comment (chrismwendt) [reply web/src/repo/blob/Blob.tsx]: I got some type errors when using `.pipe`, so I went with a mutable variable for now.

**Comment by chrismwendt** [comment]
Merging so we can start experimenting. I'm happy to continue the review process post-merge.

**Pull request merged**